salt/utils/dns.py

```diff
@@ -1,18 +1,563 @@
+    OrderedDict({
+        'prio': int,
+        'weight': int,
+        'port': to_port,
+        'name': str,
+    })
+    '10 20 25 myawesome.nl'
+
+    '''
+    try:
+        rdata = rdata.split(' ', len(rschema))
+        rschema = rschema.items()
+        return dict((
+            (fname, rcb(rdata)) for (fname, rcb), rdata in zip(rschema, rdata)
+        ))
+    except:  # pylint: disable=bare-except
+        log.error('Cant parse DNS record data: {}'.format(rdata))
+        return False
+
+
+def _rec2data(*rdata):
+    return ' '.join(rdata)
+
+
+def _to_port(port):
+    try:
+        port = int(port)
+        assert 1 <= port <= 65535
+        return port
+    except (ValueError, AssertionError):
+        raise ValueError('Invalid port {0}'.format(port))
+
+
+def _query_simple(name, rdtype, timeout=None):
+    '''
+    Use Python's socket interface to lookup addresses
+    :param name: Name of record to search
+    :param rdtype: A or AAAA
+    :param timeout: ignored
+    :return: [] of addresses or False if error
+    '''
+    sock_t = {
+        'A': socket.AF_INET,
+        'AAAA': socket.AF_INET6
+    }[rdtype]
+
+    if timeout:
+        log.warn('Ignoring timeout on simple resolver; fix resolv.conf to do that')
+
+    try:
+        addresses = [sock[4][0] for sock in socket.getaddrinfo(name, None, sock_t, 0, socket.SOCK_RAW)]
+        return addresses
+    except socket.gaierror:
+        return False
+
+
+def _query_dig(name, rdtype, timeout=None, servers=[], secure=None):
+    '''
+    Use dig to lookup addresses
+    :param name: Name of record to search
+    :param rdtype: DNS record type
+    :param timeout: server response timeout
+    :return: [] of records or False if error
+    '''
+    cmd = 'dig +search +fail +noall +answer +noclass +nottl -t {0} '.format(rdtype)
+    if servers:
+        cmd += ''.join(['@{0} '.format(srv) for srv in servers])
+    if timeout is not None:
+        if servers:
+            timeout = int(float(timeout) / len(servers))
+        else:
+            timeout = int(timeout)
+        cmd += '+time={0} '.format(timeout)
+    if secure:
+        cmd += '+dnssec +adflag '
+
+    cmd = __salt__['cmd.run_all'](cmd + str(name), python_shell=False, output_loglevel='quiet')
+
+    # In this case, 0 is not the same as False
+    if cmd['retcode'] != 0:
+        log.warning(
+            'dig returned ({0}): {1}'.format(
+                cmd['retcode'], cmd['stderr']
+        ))
+        return False
+
+    validated = False
+    res = []
+    for line in cmd['stdout'].splitlines():
+        rname, rtype, rdata = line.split(None, 2)
+        if rtype == 'CNAME' and rdtype != 'CNAME':
+            continue
+        elif rtype == 'RRSIG':
+            validated = True  # TODO: Or not
+            continue
+        res.append(rdata.strip(string.whitespace + '"'))
+
+    if secure and not validated:
+        if res:
+            return False
+        else:
+            return []
+
+    return res
+
```

Comment:
as mentionned above, a ` def _query_dig(name, rdtype, timeout=None, servers=[], secure=None):` would be interesting to add :)